fms/modules/attention.py

```diff
@@ -15,6 +16,173 @@
 from fms.modules.tp import TPModule
 
 
+class QKV(nn.Module, metaclass=abc.ABCMeta):
+    """Simple module for applying qkv in attention"""
+
+    def __init__(
+        self,
+        emb_dim: int,
+        nheads: int,
+        kvheads: int,
+        emb_kq_per_head: int,
+        emb_v_per_head: int,
+        use_bias: bool,
+        *args,
+        **kwargs,
+    ):
+        super().__init__(*args, **kwargs)
+        self.emb_dim = emb_dim
+        self.nheads = nheads
+        self.kvheads = kvheads
+        self.emb_kq_per_head = emb_kq_per_head
+        self.emb_v_per_head = emb_v_per_head
+        self.use_bias = use_bias
+
+    @abc.abstractmethod
+    def forward(
+        self, q: torch.Tensor, k: Optional[torch.Tensor], v: Optional[torch.Tensor]
+    ) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+        """applies query/key/value transformations on q, k, v inputs respectively and returns the resulting values
+
+        Args:
+            q: torch.Tensor
+                the query tensor
+            k: Optional[torch.Tensor]
+                the optional key tensor
+            v: Optional[torch.Tensor]
+                the optional value tensor
+
+        Returns:
+        Tuple[torch.Tensor, torch.Tensor, torch.Tensor]
+            the query, key, and value computed
+        """
+        pass
+
+    @abc.abstractmethod
+    def reset_parameters(self):
+        """resets the query, key, and value weights for training
+
+        Args:
+            gain: int
+                gain for std in norm (default is 1)
+        """
+        pass
+
+
+class UnfusedQKV(QKV):
+    """
+    Unfused Weights implementation of QKV
+    """
+
+    def __init__(
+        self,
+        emb_dim: int,
+        nheads: int,
+        kvheads: int,
+        emb_kq_per_head: int,
+        emb_v_per_head: int,
+        use_bias: bool,
+        *args,
+        **kwargs,
+    ):
+        super().__init__(
+            emb_dim,
+            nheads,
+            kvheads,
+            emb_kq_per_head,
+            emb_v_per_head,
+            use_bias,
+            *args,
+            **kwargs,
+        )
+        self.query = nn.Linear(
+            self.emb_dim, self.nheads * self.emb_kq_per_head, bias=use_bias
+        )
+        self.key = nn.Linear(
+            self.emb_dim, self.kvheads * self.emb_kq_per_head, bias=use_bias
+        )
+        self.value = nn.Linear(
+            self.emb_dim, self.kvheads * self.emb_v_per_head, bias=use_bias
+        )
+
+    def reset_parameters(self):
+        for m in self.modules():
+            if isinstance(m, nn.Linear):
+                nn.init.trunc_normal_(m.weight, mean=0.0, std=0.02)
+                if self.use_bias:
+                    m.bias.data.zero_()
+
+    def forward(
+        self, q: torch.Tensor, k: Optional[torch.Tensor], v: Optional[torch.Tensor]
+    ) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+        if (k is None and v is None) or (k is q and v is q):
```

Comment:
if `k is q and  is q` then why are we setting k and v to q? aren't they already set?